Bust the request cache when the mapping changes

nik9000 opened the pull request:
This makes sure that we only serve a hit from the request cache if it
was build using the same mapping and that the entire search phase is
processed using the same mapping. These are both bugs that Elasticsearch
has had for a long, long time but now that we have runtime fields they
are much more likely to matter because runtime fields can be changed
drastically from moment to moment.

Closes #62033

nik9000 committed:
Bust the request cache when the mapping changes

This makes sure that we only serve a hit from the request cache if it
was build using the same mapping and that the entire search phase is
processed using the same mapping. These are both bugs that Elasticsearch
has had for a long, long time but now that we have runtime fields they
are much more likely to matter because runtime fields can be changed
*drastically* from moment to moment.

It does by adding a counter to the local mapping that is incremented
whenever the mapping changes. This counter and the mapping is stored
under a single `volatile` variable that is updated to point to the new
mapping whenever the mapping changes. When the search phase begins we
read the contents of the `volatile` and use it to power the entire
phase, and including the version counter into the cache key.

Before this change we'd go back to the `volatile` over and over again,
potentially getting a new version of the mapping every time. This was
convenient but made it possible that we'd see the mapping update half
way through the search phase which could cause trouble.

Mechanically, this creates a `snapshot()` method on `MapperService` to
do the volatile read of the current mapping. The snapshot does all of
the reading we need from here on out. I kept all of the methods on
`MapperService` that read directly from the mapping but deprecated all
of them because using them can lead to seeing the mapping updates when
you don't expect them. We can slowly remove the usages in follow up
changes. This change is big enough as is.

Sadly, the `Mapper`s are also mutable. Luckilly, we only mutate them
before sticking them behind the `volatile` reference. Hopefully. There
is a lot of code there and it is hard to be sure. But it looks like we
*intend* to do this. We should make them properly immutable in a follow
up change.

Closes #62033

nik9000 committed:
Refactor

nik9000 committed:
Refactor again

nik9000 committed:
Javadoc

nik9000 committed:
Moar javadoc

nik9000 committed:
Merge branch 'master' into cache_bust

nik9000 committed:
Add example in core

nik9000 committed:
Merge branch 'master' into cache_bust

nik9000 committed:
WIP

nik9000 committed:
WIP

nik9000 committed:
Merge branch 'master' into cache_bust

nik9000 committed:
ooops:

nik9000 committed:
Compile plz

nik9000 committed:
Fix test

nik9000 committed:
Checkstyle

nik9000 committed:
Merge branch 'master' into cache_bust

nik9000 committed:
WIP

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Checkstyle

nik9000 committed:
Implement sourceEnabled

nik9000 committed:
Fixup test

nik9000 commented:
Draft so I can iterate on this while Jenkins runs the tests for me.

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Fix test

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 commented on modules/parent-join/src/test/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregatorTests.java:
```diff
@@ -272,34 +265,14 @@ private static SortedDocValuesField createJoinField(String parentType, String id
     }
 
-    @Override
-    protected MapperService mapperServiceMock() {
```
Moving to #66395.

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -604,9 +604,26 @@ public QueryShardContext newQueryShardContext(
+            bigArrays,
+            indexCache.bitsetFilterCache(),
+            indexFieldData::getForField,
+            mapperService(),
```
I *could* pass just the MapperService here but then we'd often have to mock the mapper service just to return the lookup.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:
```diff
@@ -115,13 +113,13 @@ private DocumentMapper(IndexSettings indexSettings,
         } catch (Exception e) {
             throw new ElasticsearchGenerationException("failed to serialize source for type [" + type + "]", e);
         }
+        this.fieldMappers = MappingLookup.fromMapping(mapping, this::parse, w -> w.write(mappingSource.compressed()));
```
The extra arguments are required so that `MappingLookup` can do all the things we need it to do now.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -399,30 +398,20 @@ public DocumentMapperForType documentMapperWithAutoCreate() {
-     */
-    public Set<String> sourcePath(String fullName) {
-        return this.mapper == null ? Collections.emptySet() : this.mapper.mappers().fieldTypes().sourcePaths(fullName);
+    public MappingLookup lookup() {
```
Probably should have javadoc.

nik9000 replied:
:+1:

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -132,7 +131,7 @@ public MapperService(IndexSettings indexSettings, IndexAnalyzers indexAnalyzers,
 
     public boolean hasNested() {
-        return this.mapper != null && this.mapper.hasNestedObjects();
+        return lookup().hasNested();
```
Instead of two volatile reads this method now does one. I think that is nice.

nik9000 commented on server/src/test/java/org/elasticsearch/index/mapper/DocumentFieldMapperTests.java:
```diff
@@ -1,139 +0,0 @@
-
-public class DocumentFieldMapperTests extends LuceneTestCase {
-
-    private static class FakeAnalyzer extends Analyzer {
```
All of this has been moved.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -288,16 +296,24 @@ protected AggregationContext createAggregationContext(IndexSearcher indexSearche
+     * {@link ObjectMapper}s to add to the lookup. By default we don't need
+     * any {@link ObjectMapper}s but testing nested objects will require adding some.
+     */
+    protected List<ObjectMapper> objectMappers() {
```
This is required by this change because we can no longer mock anything and magic nested fields into place with any object that starts with `nested_`. Now tests that use nested fields need to declare them

nik9000 replied:
We couldn't easily mock `MappingLookup` until yesterday because we exposed `FieldTypeLookup`. But now that we don't we could. I don't particularly want to though. I kind of like having a real lookup now. But maybe I've just gotten used to it.

jtibshirani replied:
Got it, it just seems like this adds some complexity to the aggregation test framework. I don't feel strongly about this though, perhaps the nested/ object mapper set-up here is already a bit complex (as seen in your comment below about `mapperService.hasNested()`).

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -444,18 +433,7 @@ public ObjectMapper getObjectMapper(String name) {
      *                                  directly associated index-time analyzer
      */
     public NamedAnalyzer indexAnalyzer(String field, Function<String, NamedAnalyzer> unindexedFieldAnalyzer) {
```
I remembered that search analyzers can be [reloaded](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-reload-analyzers.html) which doesn't cause a mapping update but changes how fields are analyzed! This is also an important way in which `MappingLookup` is not immutable.

nik9000 replied:
Yikes! We probably should figure out a way to make that replace the `MappingLookup` somehow. Do you want to save that for a follow up or for me to grab it now?

nik9000 replied:
These are required for `QueryShardContext` to get the snapshot, yeah.

nik9000 replied:
I think if we don't use the snapshot like we are here there are (rare) cases where you could see the addition of a new field without it getting in the snapshot's key. That is probably safe but its the kind of thing I don't want to have to think about and moving it over feels pretty safe too. But I think you are right about looking at this method again to make sure we actually need it.

jtibshirani replied:
I think that in the near term we should still refactor `QueryShardContext` to not use `MapperService` to retrieve index analyzers. For me it makes the approach simpler and easier to understand. I agree though that this doesn't fix the search analyzer problem tracked in #66722, which I don't feel is super high priority and could be addressed at a later time.

jtibshirani reviewed (commented):
I took a look and left some early comments. Thanks for this second iteration.

nik9000 commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -1408,7 +1408,8 @@ public void loadIntoContext(ShardSearchRequest request, SearchContext context, Q
 
         boolean[] loadedFromCache = new boolean[] { true };
-        BytesReference bytesReference = cacheShardLevelResult(context.indexShard(), directoryReader, request.cacheKey(),
+        BytesReference cacheKey = request.cacheKey(context.getQueryShardContext().mappingKey());
```
The cache key is serialized so we have to serialize *something*. We could use the hash code of the lookup but it when we talked about his in a group a week ago we said we just didn't want to have to think about the chance of collisions.

nik9000 replied:
I talked with @jtibshirani and she pointed out that we *don't* have to serialize something. She totally said that 13 hours ago but I didn't understand because I was being dense. I'm going to try and do the thing she proposed.

nik9000 replied:
And done!

nik9000 commented on server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -119,6 +123,7 @@ public QueryShardContext(
         BitsetFilterCache bitsetFilterCache,
         TriFunction<MappedFieldType, String, Supplier<SearchLookup>, IndexFieldData<?>> indexFieldDataLookup,
         MapperService mapperService,
```
We could. That require moving a couple other things to `MappingLookup` and it is the kind of thing I did in my first iteration but folks didn't like how "big" the abstraction got.

javanna replied:
I am really on the fence about this. I also would like to remove references to MapperService, yet I don't see a contained way of doing it, and I would not want to add all kinds of methods to MappingLookup. I wonder if it would be acceptable to consider this a (very important) follow-up, meaning that 7.11 will still have direct MapperService access while we are going to think about a long term solution for 7.x and master moving forward.

nik9000 replied:
I'd like to delay this simply because we aren't sure we want this. I think its reasonable but I understand the argument of wanting the `MapperService` reference for the "immutable stuff". It just that `MapperService` lets you do the mutable read too and it'd be nice not to have to worry about accidentally doing it.

If we had a separate object or interface for the immutable stuff I think it'd be really compelling to keep them separate.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Fixup

nik9000 committed:
Remove spacing change

nik9000 committed:
Javadoc

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Switch cache around

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -28,19 +29,45 @@
     private final FieldTypeLookup fieldTypeLookup;
     private final int metadataFieldCount;
     private final Map<String, NamedAnalyzer> indexAnalyzers = new HashMap<>();
+    private final Function<SourceToParse, ParsedDocument> documentParser;
```
@javanna and I talked earlier today and we thought that this function hid a volatile read. I *think* it doesn't. It is genuinely weird to be able to parse documents at search time but that is how the percolator does its thing.

jtibshirani replied:
I also don't think it hides a volatile read, but I do find the logic hard to follow: this function uses `DocumentMapper`, which is a holder object with more state. For this change I think it'd be cleanest to remove this function parameter and use `MapperService` directly to perform parsing: `mapperService.documentMapper().parse(...)`. Then we can figure out how to remove this as we work through the other follow-ups.

javanna replied:
I am also under the impression that we could do without a function, which ends up hiding the dependency on DocumentMapper, though I like that the dependency is then on a single method rather than on all of DocumentMapper or MapperService which end up exposing much more stuff that is not needed and I would not want to end up relying on.

I have been trying to simplify this by removing the need for the whole document mapper. I was not incredibly sure but I gave it a try with #66780 and I think we may be able to use DocumentParser directly here, which I think may make us all happier.

For the sake of this PR, I don't have a strong opinion but I am slightly in favour of leaving the function knowing that we have plans to replace it. If you feel strongly about `mapperService.documentMapper().parse(...)` I won't object ;)

jtibshirani replied:
I don't have strong feelings, given we are tracking plans to replace it! I had missed that when I tried to summarize the follow-ups.

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
We use up to field 10!

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -111,7 +111,7 @@ public void execute(SearchContext context) {
 
         List<FetchSubPhaseProcessor> processors = getProcessors(context.shardTarget(), fetchContext);
-        NestedDocuments nestedDocuments = context.getNestedDocuments();
+        NestedDocuments nestedDocuments = context.getQueryShardContext().getNestedDocuments();
```
This is the type of change that we could make as a follow-up to keep this one contained. If we made this separately, there would probably be other things that catch our eyes in DefaultSearchContext to address with it. Also, we are not after fixing the fetch phase, so what are the gains? Could we leave the method in SearchContext for now and retrieve the MappingLookup from the MapperService in this case, and file it as a followup?

javanna replied:
Along the same lines, I wonder: if we avoid moving the method to QueryShardContext, do we still need to migrate NestedDocuments to MappingLookup, and to consequently move the method it uses from DocumentMapper to MappingLookup? It is a good change I think, but I don't see if it's required. In fact, QueryShardContext only needs hasNested which requires no changes as it's already part of MappingLookup? Am I missing anything?

nik9000 replied:
Makes sense to me. I'll have another look. I remember it being obvious to me at the time but I think all kinds of silly things.

nik9000 replied:
I've removed this change and added it as a follow up listed in #64663.

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/MappingLookupTests.java:
```diff
@@ -53,20 +63,135 @@ public void testSubfieldOverride() {
+            }).tokenStream("field3", "blah"));
+    }
+
+    private void assertAnalyzes(Analyzer analyzer, String field, String output) throws IOException {
```
s/Analyzes/Analyzers

nik9000 replied:
I think it is right as is. It is asserting that the term analyzes to the output term.

javanna reviewed (commented):
I left a first batch of comments. I plan on taking a second look and dig deeper into a couple of aspects that are not completely clear to me. Though I like the approach, I think it's the right direction.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:
```diff
@@ -105,7 +103,7 @@ public DocumentMapper build() {
     private final MetadataFieldMapper[] noopTombstoneMetadataFieldMappers;
 
-    private DocumentMapper(IndexSettings indexSettings,
+    protected DocumentMapper(IndexSettings indexSettings,
```
Yeah, I think this is a leftover.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:
```diff
@@ -115,13 +113,14 @@ private DocumentMapper(IndexSettings indexSettings,
         } catch (Exception e) {
             throw new ElasticsearchGenerationException("failed to serialize source for type [" + type + "]", e);
         }
+        // TODO this::parse performs a volatile read on mapping from MapperService. Yikes!
```
Yeah. I was just realizing last week that we're ok here. I'll drop it. It is pretty tricky though!

javanna replied:
Agreed, I have some plans to maybe make it a bit better, but still, Mapping and some other stuff is needed all the way in ParseContext. I was wondering if parse does fit in MappingLookup, and I think the answer is that we don't have many alternatives. The only one I can think of is forget about percolator and let it do a volatile read, but it would be a shame in the context of this whole PR, so I think we are good.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:
```diff
@@ -124,7 +133,10 @@ MappedFieldType get(String field) {
      * @return A set of paths in the _source that contain the field's values.
      */
-    Set<String> sourcePaths(String field) {
+    public Set<String> sourcePaths(String field) {
```
I'm ok either way.

javanna replied:
I think that for instead simpleMatchtoFullName is already exposed that way, so let's do it that way and restore the original visibility of FieldTypeLookup?

javanna replied:
sorry I thought it was a typo

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -234,4 +274,116 @@ private static String parentObject(String field) {
+    /**
+     * Returns all nested object mappers which contain further nested object mappers
+     *
+     * Used by BitSetProducerWarmer
```
:+1:

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -234,4 +274,116 @@ private static String parentObject(String field) {
+    }
+
+    public boolean isEmpty() {
+        return objectMappers.isEmpty() && fieldTypeLookup.isEmpty();
```
I think if we go that route we'd want some protection to make sure that any code the tries to build an empty mapping would get `EMPTY`. It doesn't feel worth it to me but I could do it if you thought it was imporant.

javanna replied:
I don't follow: hasMappings was checking if document mapper is null. You are now returning EMPTY whenever document mapper is null. What else is needed? Maybe bad naming is the cause of misunderstandings here. To me it is not about empty mappings, but rather no mappings or any mappings.

nik9000 replied:
Yeah, maybe its bad naming. If I renamed it to, like, `hasMappings` or something then I think it'd be cool to check for EMPTY.

javanna replied:
cool, the method that calls it is named hasMappings, in QueryShardContext

javanna replied:
This may be something different from the other internal usage you have introduced in FieldTypeLookup, but that one does not need a public method.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -28,19 +29,45 @@
+    /**
+     * Key for the lookup to be used in caches.
+     */
+    public class CacheKey {
```
:+1:

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -28,19 +29,45 @@
 import java.util.stream.Stream;
 
-public final class MappingLookup {
+public class MappingLookup {
```
:+1:

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -132,7 +131,7 @@ public MapperService(IndexSettings indexSettings, IndexAnalyzers indexAnalyzers,
     }
 
     public boolean hasNested() {
```
Oh nice!

nik9000 replied:
I did that sort of thing in my original cache busting change and it made the change too difficult to review. I'd like to save it for a follow up so we can get this one in.

jtibshirani replied:
I'm in agreement!

javanna replied:
I tried to clarify this in the meta issue: one thing is to care for consumers accessing the same snapshot, which sounds to me like a big project. Another thing is, I think, to remove duplicated methods that are exposed in multiple places (e.g. `simpleMatchToFullName`) and could be rather called directly retrieving `MappingLookup`

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Fixup

nik9000 committed:
Iter

nik9000 committed:
link to public

nik9000 committed:
Revert cleanup of nested

nik9000 committed:
Fix test

nik9000 committed:
Finish reverting nested change

nik9000 committed:
Checkstyle, you loveable scamp

nik9000 committed:
Probably not neeeded

nik9000 commented:
@javanna I believe I've addressed all of your comments. I reverted the `NestedDocuments` change which shrunk this some. And, you are right, we'll want to think more about it when we do it. I think I picked it up as part of this originally as left over from the first iteration.

nik9000 committed:
Revert moar

nik9000 committed:
Document not fixed issue

jtibshirani commented on server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java:
```diff
@@ -128,6 +129,15 @@ BytesReference getOrCompute(CacheEntity cacheEntity, CheckedSupplier<BytesRefere
                 }
             }
+            /*
+             * note that we don't use a closed listener for the mapping. Instead
```
Thanks for adding this clarifying comment.

jtibshirani reviewed (commented):
I added some small code-level comments. It's looking quite close to me! I had two remaining high-level questions:
* Will we try to remove `MapperService` from `QueryShardContext`? It feels like a solid and clear approach: we'd know search could only access `MappingLookup`, which is designed to contain the latest mapping information. I didn't get the chance to try out the required refactors, but I will take out a look if you agree it's a useful goal.
* Related to this, it'd be good to remove references to `MapperService` inside the `SearchContext` implementations. That way we can be sure that search always accesses mapping information through `QueryShardContext`, instead of circumventing it. Maybe we could track this as a follow-up?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -399,30 +398,23 @@ public DocumentMapperForType documentMapperWithAutoCreate() {
-        if (fullName.equals(TypeFieldType.NAME)) {
-            return new TypeFieldType(this.mapper == null ? "_doc" : this.mapper.type());
-        }
-        return this.mapper == null ? null : this.mapper.mappers().fieldTypes().get(fullName);
```
let's track this as a follow-up so we can keep this change as contained as possible?

nik9000 replied:
Given how folks received my last attempt at this I wanted to keep my change as small as I could manage. I think it is correct to switch them all to `mappingLookup` and my instinct is to do it now, while I'm here. But I don't want to make this harder to review.

javanna replied:
I think that would not only make it harder to review, but it would make the PR bigger by making changes that are not strictly necessary. I think we should totally do it, but as a follow-up which is not going to be backported to 7.11

jtibshirani replied:
It works for me to do it in a follow-up.

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java:
```diff
@@ -46,6 +47,17 @@ public void testPreflightUpdateDoesNotChangeMapping() throws Throwable {
         assertThat("field was not created by mapping update", mapperService.fieldType("field0"), notNullValue());
     }
 
+    public void testLookup() throws IOException {
```
testMappingLookup?

nik9000 replied:
7b7cc2786da39c1e5e81435cbec1b5ed4cb15b07

javanna commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MappingLookupUtils.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public static FieldMapper mockFieldMapper(MappedFieldType type) {
+        FieldMapper mapper = mock(FieldMapper.class);
```
++

nik9000 replied:
65962e2fdd083fb54ac8ef53425216475cecc0dc

javanna commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MappingLookupUtils.java:
```diff
@@ -0,0 +1,67 @@
+        );
+    }
+
+    public static FieldMapper mockFieldMapper(MappedFieldType type) {
```
The way that this is used looks suspicious to me. I see it being used with runtime fields, which don't have a corresponding mapper normally. With this method they end up being exposed as ordinary fields to tests?

javanna replied:
There should be a separate code-path for creating runtime fields, that only provides them to MappingLookup without them causing the creation of field mappers.

nik9000 replied:
`QueryShardContextTests` has been doing that sort of thing for a month or so now. We didn't have runtime fields without the `MappedFieldType` back then. I've pushed 83f9f30acfe3557643603d00444fd05ff48a1c42 to fork off runtime fields. I'm not a huge fan of the `instanceof` that it adds but it seems like a useful thing for testing.

javanna commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -227,38 +225,41 @@ protected AggregationContext createAggregationContext(IndexSearcher indexSearche
-            searchLookup) -> fieldType.fielddataBuilder(mapperService.getIndexSettings().getIndex().getName(), searchLookup)
+            searchLookup) -> fieldType.fielddataBuilder(indexSettings.getIndex().getName(), searchLookup)
                 .build(new IndexFieldDataCache.None(), breakerService);
+        BitsetFilterCache bitsetFilterCache = new BitsetFilterCache(indexSettings, new BitsetFilterCache.Listener() {
```
how bad would you feel about leaving this test as-is, and simply providing mapperService.mappingLookup() to the query shard context constructor? Would that work? I figured we'll get back to this anyways once we work on NestedDocuments access.

nik9000 replied:
I didn't mock `MappingLookup` at first because we were exposing `FieldTypeLookup` and it would have made mocks returning mocks which would have been a pain. So I ended up doing this to use the "real" stuff because it was just a ton of work to mock.

I guess at this point we don't need to keep them because we are routing everything through `MappingLookup` but they were a pain to get right so I'd like to keep them. But if you are really uncomfortable with them I could try and fork them out and get them back in.

nik9000 replied:
I couldn't keep the test as is though - I'd have to mock `MappingLookup` instead of `MapperService` i believe.

javanna commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexReaderWrapperIntegrationTests.java:
```diff
@@ -182,15 +173,12 @@ protected IndicesAccessControl getIndicesAccessControl() {
+            new KeywordFieldType("field"),
+            new KeywordFieldType("f1"),
+            new KeywordFieldType("f2")
+        );
```
these changes are great. They make what tests do much more readable, and also allow us to only plug in the needed behaviour. Before the field type lookup would always return the same field type regardless of the provided field name, which is not so easy to follow.

javanna commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -325,11 +333,13 @@ public boolean shouldCache(Query query) {
         when(indexShard.shardId()).thenReturn(new ShardId("test", "test", 0));
         when(ctx.indexShard()).thenReturn(indexShard);
+        MapperService mapperService = mock(MapperService.class);
+        when(mapperService.hasNested()).thenReturn(false);
```
This one throws me off: why do we need to mock hasNested now compared to before? It looks to me that these test changes are not strictly required and could be made as a follow-up. But I am probably missing something.

nik9000 replied:
This is because the agg tests never want to try and fetch nested hits but are quite happy to do the aggs on them. Its *weird* but this funny dance keeps them working the way they were before.

nik9000 replied:
Previously we bumped into this behavior because of the defaults of mockito - it returns `false` unless you tell it otherwise. So I'm making this explicit....

javanna replied:
I see makes sense

javanna commented on server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java:
```diff
@@ -236,7 +248,8 @@ public boolean equals(Object o) {
             if (o == null || getClass() != o.getClass()) return false;
             Key key = (Key) o;
-            if (Objects.equals(readerCacheKey, key.readerCacheKey) == false) return false;
+            if (mappingCacheKey.equals(key.mappingCacheKey) == false) return false;
```
shouldn't we add mappingCacheKey to hashcode as well? Maybe we should expand IndicesRequestCacheTests so that it catches this?

nik9000 replied:
I'm not sure how I managed to miss that. I'll do it.

nik9000 replied:
78d5b334f2526363f58dc35fd343239bd3225c48

javanna commented on server/src/test/java/org/elasticsearch/indices/IndicesRequestCacheTests.java:
```diff
@@ -198,13 +203,93 @@ public void testCacheDifferentReaders() throws Exception {
+    public void testCacheDifferentMapping() throws Exception {
+        IndicesRequestCache cache = new IndicesRequestCache(Settings.EMPTY);
+        MappingLookup.CacheKey mappingKey1 = MappingLookupUtils.fromTypes().cacheKey();
+        MappingLookup.CacheKey mappingKey2 = MappingLookupUtils.fromTypes().cacheKey();
```
could we simply use an empty mapping lookup here instead of calling fromTypes without providing any type?

nik9000 replied:
We could use it for one of the keys I guess but I wanted distinct cache keys for both keys and this manages it.

javanna replied:
oh boy you are right.

javanna commented on server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -603,15 +613,19 @@ public BigArrays bigArrays() {  // TODO this is only used in agg land, maybe rem
+     * Cache key for current mapping.
+     */
+    public MappingLookup.CacheKey mappingCacheKey() {
+        return mappingLookup.cacheKey();
```
maybe I only need to get used to this approach, but I wonder if it is necessary to carry around this additional cache key. I guess it is no option to make it become part of the existing cacheKey that the request exposes or something along these lines?

nik9000 replied:
I have to carry it around somewhere for us be able to assert on it. Originally I was using a `Writeable` key and inserting it into the request's cache key. When I was just writing a long it was fine but writing the whole mapping was quite big. And it was incomplete because we couldn't possibly handle analyzer reloading. Using the dummy object is nicer because it is smaller and can handle the analyzer reloading.

But, yeah, there is a fair bit of plumbing to get it to a better spot. What were you thinking?

javanna replied:
I will defer to @jtibshirani on this. I don't have any specific idea, just noticing that while this approach has advantages, it does require some additional plumbing and complexity. Maybe it's the right compromise though. I will drop another question: do we need to worry about hash collisions?

nik9000 replied:
> I will drop another question: do we need to worry about hash collisions?

When we last talked with @jpountz he said something along the lines of "I don't want to ever has to double check that we're not getting a hash collision when helping someone debug something." The resonated with me. Hash collisions aren't likely at all but it gives us a nice piece of mind at a fairly low cost. We'd have to plumb *something* through anyway, I think. Going this way does require all those changes to the `IndicesRequestCache` but they weren't too bad. Probably worth the peace of mind.

jtibshirani replied:
I liked how this matched our approach for incorporating `DirectoryReader` into the cache key. We use a dedicated object reference, instead of keying on the actual content of `MappingLookup`. Specifically, we're not keying on a hash, so we shouldn't need to handle hash collisions ourself.

I suggested this approach as an option, but not an expert in this caching code and would be happy if you wanted to check with someone more familiar!

javanna reviewed (commented):
I did another round and left another batch of comments. Getting close indeeed!

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -28,19 +28,49 @@
 
-public final class MappingLookup {
+/**
+ * A (mostly) immutable snapshot current mapping of an index with access
```
67364828c42

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -309,7 +309,7 @@ protected void doXContentBody(XContentBuilder builder, boolean includeDefaults,
     protected abstract String contentType();
 
-    public final Map<String, NamedAnalyzer> indexAnalyzers() {
+    public Map<String, NamedAnalyzer> indexAnalyzers() {
```
I needed it to handle the mocking. It looks like you suggested in another spot to use `MockFieldMapper` which should let me avoid it.

nik9000 replied:
65962e2fdd083fb54ac8ef53425216475cecc0dc

nik9000 committed:
words are hard

nik9000 commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -227,38 +225,41 @@ protected AggregationContext createAggregationContext(IndexSearcher indexSearche
+                .collect(toList()),
+            List.of(),
+            0,
+            souceToParse -> null,
```
0f4adcca0e78c0889b172a9d91d6a4ee838157bb

nik9000 committed:
Switch to MockFieldMapper

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -504,6 +485,7 @@ public boolean isMetadataField(String field) {
                 reloadedAnalyzers.add(analyzerName);
             }
         }
+        // TODO this should make a new MappingLookup to bust the cache
```
I kind of like having the TODO just for folks reading the code to bump into the issue. I've added the issue url to it.

nik9000 committed:
Speeling

nik9000 committed:
Add link in todo

nik9000 committed:
Finish rename

nik9000 committed:
Special handling of runtime fields

nik9000 committed:
Extra casting

nik9000 committed:
equals and hashcode for the key

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -28,19 +28,49 @@
+        List.of(),
+        List.of(),
+        0,
+        souceToParse -> null,
```
d4795914b6dc5f888466e583ae22c0c85771c203

nik9000 committed:
Speeling

nik9000 commented on server/src/test/java/org/elasticsearch/indices/IndicesRequestCacheTests.java:
```diff
@@ -80,14 +85,14 @@ public void testBasicOperationsCache() throws Exception {
         assertTrue(loader.loadedFromCache);
         assertEquals(1, cache.count());
-        assertTrue(requestCacheStats.stats().getMemorySize().bytesAsInt() > value.length());
+        assertTrue(requestCacheStats.stats().getMemorySize().getBytes() > value.length());
```
This isn't strictly required but the old method is deprecated and I figured I'd clean it up while I was here.

nik9000 committed:
Start your paragraph with a capital letter

nik9000 committed:
Not null

nik9000 reviewed (commented):
Hey folks! I've pushed a bunch of updates and replied to a bunch of comments. I "resolved" all of the comments that I think my changes resolved. Please double check they all make sense to you.

I think the major open questions now are:
1. The `MapperService` reference on `QueryShardContext`
2. The test changes that I had to make before we routed everything through `MappingLookup` but now don't have to make. But I've grown used to them.
3. What exactly are the follow ups? We know we have #66722 and reviving the changes that I made to `NestedDocuments`. Are there others?

nik9000 commented:
elasticsearch-ci/1 failure was #66571.
run elasticsearch-ci/1

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 commented:
run elasticsearch-ci/1

javanna commented on server/src/test/java/org/elasticsearch/indices/IndicesRequestCacheTests.java:
```diff
@@ -441,19 +532,51 @@ public void testEqualsKey() throws IOException {
+            for (IndicesRequestCache.Key key2 : keys) {
+                if (key1 == key2) {
+                    assertEquals(key1, key2);
+                    assertEquals(key1.hashCode(), key2.hashCode());
```
we have EqualsHashCodeTestUtils, not sure it's worth using it here

nik9000 replied:
Me neither given that we already had all this stuff to test it.

javanna commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MappingLookupUtils.java:
```diff
@@ -0,0 +1,54 @@
+                runtimeFields.add((RuntimeFieldType) type);
+            } else {
+                mappers.add(mockFieldMapper(type));
+            }
```
not a fan of this `if` either. I would rather do one of the following two things:

1) subclass MappingLookup and override only the needed methods, that way we don't have to create a field mapper at all. not sure if till work everywhere, but it's what were previously doing in QueryShardContextTests.

2) have a different method that accepts RuntimeFieldTypes. Along these lines I was wondering if the following method needs to be public, and if the analyzers logic is needed in all the callers or should be rather be pushed to the callers that need it, in which case we could make callers provide field mappers instead of mapped field types as they'd only need to do  `new MockFieldMapper(fieldType)`.

nik9000 replied:
I think it'd be pretty fair to move the analyzers change into `MockFieldMapper`. It is *weird* if the analyzers don't line up with the `MappedFieldType`'s analyzers. Sort of a "testing is weird" thing.

I don't much like the subclassing idea because `MappedFieldType` does a lot of stuff when we build it and I kind of think we should let it keep doing it. I'd be all for simplifying that stuff, I guess, but as it stands now I like using it raw. I'm quite happy pass two lists to this method.

nik9000 replied:
Doing the two lists works well.

I'm going to investigate who needs the analyzers. Its weird not to have any analyzers if the field type you are using has some. But it is kind of also weird to copy the ones from the search analyzer.

nik9000 replied:
I think I've pushed a working version that isolates the analyzer bit to just the test that needs it.

javanna reviewed (approved):
I did another round and left a few other minors, mostly around tests. This looks good on my end, I will defer to @jtibshirani for the final approval.

I do care that we list the mentioned follow-up work in a meta issue so that it doesn't get lost. Some of those tasks are quite important and we will want to work on them soon. There were more compared to the two you mentioned @nik9000 could you take care of collecting them from the feedback you got?

nik9000 commented on docs/reference/indices/apis/reload-analyzers.asciidoc:
```diff
@@ -13,9 +13,17 @@ stream's backing indices.
 --------------------------------------------------
 // TEST[setup:my_index]
 
+IMPORTANT: After reloading the search analyzers you should clear the request
```
:+1:

jtibshirani replied:
Piggy-backing off this comment: it'd be great to update the "Cache invalidation" section in `request_cache.asciidoc` with a line about mapping updates.

jtibshirani replied:
Piggy-backing off this comment: it'd be great to update the "Cache invalidation" section in `request_cache.asciidoc` with a line about mapping updates.

nik9000 replied:
27eef8b884a127a2ccfc0f0e9ff881186e608354

nik9000 commented on docs/reference/indices/apis/reload-analyzers.asciidoc:
```diff
@@ -13,9 +13,17 @@ stream's backing indices.
 // TEST[setup:my_index]
 
+IMPORTANT: After reloading the search analyzers you should clear the request
+           to make sure it doesn't contain responses derrived from the
```
:+1:

nik9000 replied:
629f510ea9db1212df07531509b098766703bc27

nik9000 commented on server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java:
```diff
@@ -236,7 +248,8 @@ public boolean equals(Object o) {
             Key key = (Key) o;
-            if (Objects.equals(readerCacheKey, key.readerCacheKey) == false) return false;
+            if (mappingCacheKey.equals(key.mappingCacheKey) == false) return false;
+            if (readerCacheKey.equals(key.readerCacheKey) == false) return false;
```
I mean, I had to add the line above it and we didn't need null handling. So, no I could revert it, but it feels nice to have the two lines "line up". I could use `Objects.equals` in the new code too, but the but the `mappingCacheKey` can't be null so it'd confuse me when I go to read it later.

nik9000 committed:
Try this

jtibshirani commented:
@nik9000 and I caught up and decided to track these important follow-ups to solidify the design:
* Remove `MapperService` from `QueryShardContext`, to prevent access to inconsistent mapping information. This may involve more than one change, as it's used in a couple different ways.
* Remove an unnecessary `MapperService` reference from `SearchContext`. Search phases reference `SearchContext` in addition to `QueryShardContext`, so this is also important in making sure they don't access inconsistent mapping information.

All follow-ups we discussed are being tracked in #64663. (Although maybe they could live in a dedicated issue?)

jtibshirani reviewed (commented):
I left two last comments, other than that this looks good to me. I agree with @javanna that the follow-ups are critical!

nik9000 committed:
Merge branch 'master' into cache_bust_take_two

nik9000 committed:
Update cache invalidation

nik9000 committed:
Speeling

nik9000 commented:
run elasticsearch-ci/docs

javanna commented on docs/reference/modules/indices/request_cache.asciidoc:
```diff
@@ -32,14 +32,14 @@ Scripted queries that use the API calls which are non-deterministic, such as
 The longer the refresh interval, the longer that cached entries will remain
-valid. If the cache is full, the least recently used cache keys will be
-evicted.
+valid even if ther eare changes to the documents. If the cache is full, the
```
s/ther eare/there are

javanna commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MappingLookupUtils.java:
```diff
@@ -0,0 +1,36 @@
+    public static MappingLookup fromTypes(List<MappedFieldType> concreteFields, List<RuntimeFieldType> runtimeFields) {
+        List<FieldMapper> mappers = concreteFields.stream().map(MockFieldMapper::new).collect(toList());
+        return new MappingLookup("_doc", mappers, List.of(), List.of(), runtimeFields, 0, souceToParse -> null, true);
+    }
```
the size of this class makes me now wonder if we should drop it in favour of an additional test constructor/method in MappingLookup.

nik9000 replied:
Yeah. Me too. It's pretty small. It does depend on `MockFieldMapper` which is in the test framework. I could make move the methods to static methods on `MockFieldMapper`. It isn't *quite* right for runtime fields though.

nik9000 committed:
Typo

nik9000 commented:
>  ** [jtibshirani](/jtibshirani) ** approved these changes [ 5 minutes ago ](#pullrequestreview-558136407)

OK! As soon as Jenkins approves these I'll merge them. I've got a backport started cooking over at #66795.

jtibshirani commented:
I just realized: maybe we should update the PR title to help future readers.

nik9000 commented:
> @nik9000 I just realized: maybe we should update the PR title to help future readers.

Done!

jtibshirani commented:
The refactor #66877 made me notice that we aren't guaranteed to use the same `MappingLookup` for the query and fetch phases. This is because we create a new `QueryShardContext` in each phase. Here's my understanding of the current situation:
1. When the mapping changes, we will never serve a cached response that used an old version of the mappings.
2. We try hard to use the same mappings/ settings through the search request. However there are some cases where that's not true -- for example we could use a newer mapping version when fetching vs. querying.

It's great we covered 1, as it was a critical issue with runtime fields. I think point 2 is a strong goal, but not a blocker. Does this match your understanding too? We could file a dedicated issue to track it, listing the places where it's not yet consistent.

nik9000 commented:
Yeah. We don't do anything in particular to protect against `can_match` getting a different mapping either. But the things that can cause it to return `true` or `false` are mostly immutable at the moment.
